### PR TITLE
B 18823 ghc create requested office user int

### PR DESCRIPTION
**Pull request by cameroncaci**
## Summary

Introduces new GHC V1 endpoint `POST requested-office-users` to allow for office users to create new office users.
Introduces new service object for roles to be associated on user creation.
Updated Daniel’s create office user endpoint to also return roles

Previous PR #12180 

## Verification Steps for Reviewers

These are to be checked by a reviewer.

- [ ] Has the branch been pulled in and checked out?
- [ ] Have the BL acceptance criteria been met for this change?
- [ ] Was the CircleCI build successful?
- [ ] Has the code been reviewed from a standards and best practices point of view?

### Setup to Run the Code

- [Instructions for starting storybook](https://transcom.github.io/mymove-docs/docs/frontend/setup/storybook)
- [Instructions for starting the MilMove application](https://transcom.github.io/mymove-docs/docs/getting-started/application-setup/)
- [Instructions for running tests](https://transcom.github.io/mymove-docs/docs/getting-started/development/testing)

### How to test

1. Import latest ghc.yaml into postman
2. Grab a transportation office id from the transportation-offices GHC endpoint. Grab the “Id”, not the “Address -> Id”. There are 2 Ids per response back, grab the root one.
3. Login to office as any office user
4. Open network tab in browser dev tools and refresh page
5. Select latest network request, select headers
6. Grab ‘X-Csrf-Token’ value from the headers
7. Open postman
8. Navigate to MilMove GHC API requested-office-users “POST Create an Office User”
9. Navigate to Headers and add a new header, ‘X-Csrf-Token’, past your value in the value box
10. Update the URL to be respective of your environment, such as `officelocal:3000/ghc/v1/requested-office-users`
11. Update the body to be something like this (Update the transportationOfficeId:
```
{
  "firstName": "John",
  "lastName": "Doe",
  "email": "johndoe12@example.com",
  "telephone": "555-555-5555",
  "transportationOfficeId": "5a3388e1-6d46-4639-ac8f-a8937dc26938",
  "roles": [
    {
      "name": "Transportation Ordering Officer",
      "roleType": "transportation_ordering_officer"
    }
  ],
  "edipi": “1234456789",
  "otherUniqueId": "2234567892"
}
```
12. You should receive a return such as this
```
	{
    "active": false,
    "createdAt": "2024-03-06T17:12:41.777Z",
    "edipi": "1234456789",
    "email": "johndoe12@example.com",
    "firstName": "John",
    "id": "a2f04463-a7f7-4f8d-969f-590540c2ba13",
    "lastName": "Doe",
    "middleInitials": null,
    "otherUniqueId": "2234567893",
    "rejectionReason": null,
    "roles": [
        {
            "createdAt": "2024-03-04T13:22:01.338Z",
            "id": "ed095ebb-7291-4276-97df-3672df26b4ab",
            "roleName": "Transportation Ordering Officer",
            "roleType": "transportation_ordering_officer",
            "updatedAt": "2024-03-04T13:22:01.338Z"
        }
    ],
    "status": "REQUESTED",
    "telephone": "555-555-5555",
    "transportationOfficeId": "5a3388e1-6d46-4639-ac8f-a8937dc26938",
    "updatedAt": "2024-03-06T17:12:41.777Z",
    "userId": "1eea8c6c-3d35-454d-90e8-444fe95a205d"
}
```
13. Modify JSON body accordingly for test cases of validation

### CODE REVIEW

**Comment by danieljordan-caci** [comment]
I am getting an `Unauthorized` return when testing this.

This will need to be done before authentication. We'll need to remove the middleware check for this API endpoint, right?

**Comment by cameroncaci** [comment]
> I am getting an `Unauthorized` return when testing this.
> 
> This will need to be done before authentication. We'll need to remove the middleware check for this API endpoint, right?

What do you mean?

File: swagger-def/ghc.yaml [hunk 1]
```diff
@@ -27,6 +27,39 @@ tags:
...
 paths:
+  /requested-office-users:
+      post:
+        consumes:
+          - application/json
+        produces:
+          - application/json
+        summary: Create an Office User
+        description: >
+          Request the creation of and office user. An administrator will need to approve them after creation. Note on requirements:
```
[the remark this answers is not in the document]
Comment (cameroncaci) [reply swagger-def/ghc.yaml:38]: Thanks for the catch, resolved now

**Comment by danieljordan-caci** [comment]
> > I am getting an `Unauthorized` return when testing this.
> > This will need to be done before authentication. We'll need to remove the middleware check for this API endpoint, right?
> 
> What do you mean?

Requesting office users will need to use this endpoint prior to authentication and will not be logged into MilMove when using the form that will call this API endpoint so we'll need to figure out how to remove the authentication requirement (which I'm assuming is some middleware) to use it.

**Comment by cameroncaci** [comment]
@danieljordan-caci Are we sure the GHC endpoint isn't for Office users (Like the service counselor) to request new accounts? Is there a separate endpoint for public account requesting?

**Comment by danieljordan-caci** [comment]
> @danieljordan-caci Putting this PR back up for review. As the ticket has no explicit mention of the no-auth requirement, this ticket has been completed per its requirements. Additional requirements should involve ticket enhancement / new backlog item creation (Which I've pinged Danny for). Once we get these confirmed enhancements we can modify this ticket or create a new backlog item

A known requirement is that this needs to be done outside of authentication (per the demo that was approved by Kipp). I understand that the endpoint is functional when logged in and authenticated, but does not pass practical testing on my end (trying to use the endpoint outside of authentication).

I'll wait to see what Danny says in Agility before approving, but right now it's an endpoint that can not be used for its intended functional purpose.

**Comment by cameroncaci** [comment]
> > @danieljordan-caci Putting this PR back up for review. As the ticket has no explicit mention of the no-auth requirement, this ticket has been completed per its requirements. Additional requirements should involve ticket enhancement / new backlog item creation (Which I've pinged Danny for). Once we get these confirmed enhancements we can modify this ticket or create a new backlog item
> 
> A known requirement is that this needs to be done outside of authentication (per the demo that was approved by Kipp). I understand that the endpoint is functional when logged in and authenticated, but does not pass practical testing on my end (trying to use the endpoint outside of authentication).
> 
> I'll wait to see what Danny says in Agility before approving, but right now it's an endpoint that can not be used for its intended functional purpose.

Understood on that part, that means this ticket was written improperly. It's not a jab at you, it's that tickets need to be written so that upon completing and meeting AC criteria, it is complete. And it must be written so that anyone from any team can take and action. This is apparently not the case for this ticket, meaning it is faulty.

**Review by traskowskycaci — APPROVED** [review]
Works as expected! If there is more work to be done on this branch as an outcome of the discussions on this story, let me know and I will review again.

**Review by TevinAdams — APPROVED** [review]
Tested and reviewed.

**Comment by cameroncaci** [comment]
@TevinAdams @traskowskycaci re-requesting review as the `requested-office-users` endpoint has been opened to be publicly accessible

**Review by traskowskycaci — APPROVED** [review]
Retested, looks good

**Review by danieljordan-caci — APPROVED** [review]
Works perfectly! Nice job on figuring out this. LGTM!

**Pull request merged**